### ISSUE TITLE
Consider allowing default in quantile and median

### ISSUE THREAD

**Issue by bkamins** [JuliaStats/Statistics.jl#132]
In generic code the fact that the following operations error:

```
julia> median(Int[])
ERROR: ArgumentError: median of an empty array is undefined, Int64[]

julia> quantile(Int[], 0.5)
ERROR: ArgumentError: empty data vector
```

is inconvenient (as the only work-around is try-catch). Maybe we could allow passing `default` kwarg that would be returned when the passed collection is empty?

**Comment by bkamins** [comment]
> Currently, empty mean returns NaN for numbers and throws for non-numbers.

Yes, but at least it does not throw an error in standard case (but indeed we could allow for providing the other value of default as `NaN`, while correct is highly non-standard for Julia data processing)

> called `init` there

The reason is that `sum` is a reduction, while e.g. `median` is not a reduction, so that is why `init` is not a best name in this case. But indeed - I was not sure what would be the best name of the kwarg.

**Comment by pdeffebach** [comment]
FWIW, R returns `NULL` which is roughly equivalent to `nothing`.

**Comment by bkamins** [comment]
These are all good ideas. But then the question is how to convert it to API in a non-breaking way?
We probably need a kwarg that would allow to switch behavior.

**Comment by bkamins** [comment]
> Seems nonbreaking, and allows for easy default selection through something(...).

It would be non-breaking, but I feared proposing it for such a common function 😄.

>  a general wrapper function so that the interface looks like `onempty(median, xs, default=0.0)`.

This general wrapper function would require special casing for each function passed as its first argument to be efficient.

The point is that `isempty(skipmissing(x))` is O(n) operation:

```
julia> x = missings(Int, 10^8);

julia> @time isempty(skipmissing(x))
  0.077613 seconds (1 allocation: 16 bytes)
true
```

But maybe it is good enough?

**Comment by bkamins** [comment]
@nalimilan - what do you think?

**Comment by nalimilan** [comment]
I don't think there's a precedent of a function that returns `nothing` when the input is empty, right? So adding a `default` keyword argument seems simpler. It's too bad we can't call it `init` for consistency with `sum`.

**Comment by bkamins** [comment]
In general where possible I personally prefer special value-based behavior rather than error-based. For `median` and `quantile` returning `nothing` is safe.

However, I understand the concern of @nalimilan that it requires a careful design so that we have one common solution rather than many special cases. E.g. `nothing` as a special value works for `mean`, `median`, `quantile`, and e.g. `std`, but it is not a good value for `sum` as:
```
julia> isnothing(sum([nothing]))
true
```

**Comment by bkamins** [comment]
I was thinking what is best and came to the conclusion that we could define somewhere:
```
withempty(f, default) = x -> isempty(x) ? default : f(x)
```
(name is tentative)

With `skipmissing` the performance will not be optimal, but I think most of the time users will not complain (you would need to have a vector of all `missing` values with millions of such elements to notice the problem.

In [this post](https://bkamins.github.io/julialang/2022/12/17/empty.html) I have shown an example usage.

**Comment by bkamins** [comment]
We could define:
```
withcondition(f; default=missing, condition=isempty) = x -> condition(x) ? default : f(x)
```
(I proposed adding defaults, but we could skip them).

Then, by default we would have "return `missing` if empty, but one could write `withcondition(median, default=NaN, condition=x->length(x) < 5)` to get what you want.

**Comment by nalimilan** [comment]
That sounds overkilll to me. The point of such a function is to be simple and convenient, as users can easily define other functions if needed. Actually we could almost have `emptymissing(f) = x -> isempty(x) ? missing : f(x)` to make things even simpler for the common case (and add that to Missings.jl).

**Comment by bkamins** [comment]
I would be also OK with such a solution as it likely covers most of the use cases.

**Comment by jariji** [comment]
After reading @bkamins blog post I came here to suggest

```
whenever(predicate, f, default) = x -> predicate(x) ? default : f(x)
```

It works for `isempty`, `isnothing`, `ismissing`, `isnan`. I like the positional version more than the kwargs version suggested above. I'm not committed to the name at the moment.

**Comment by nalimilan** [comment]
`whenever(predicate, f, default)` is only 3 characters shorter than `x -> predicate(x) ? default : f(x)` so I don't really see the point of adding such a general function.

**Comment by bkamins** [comment]
> I don't really see the point of adding such a general function.

There are two points that can be considered:
1. you actually need to write `(x -> predicate(x) ? default : f(x))` in practice, and these extra `(` and `)` create visual noise and are easy to forget.
2. `x -> predicate(x) ? default : f(x)` will be re-compiled every time it is written, while `whenever(predicate, f, default)` is compiled once

**Comment by jariji** [comment]
BTW, I'm not totally sold on the argument order being `f, default`. `whenever(predicate, value, function)` seems closer to the familiar `if-then-else` ordering. 

Tangential, but I don't really like that the predicate functions we have are positive in the degenerate case rather than the good case. If there were better words for `nonmissing`/`nonempty`/`nonnan` that could be nicer than the current system.

**Comment by mreichMPI-BGC** [comment]
> We could define:
> `withcondition(f; default=missing, condition=isempty) = x -> condition(x) ? default : f(x)`
> (I proposed adding defaults, but we could skip them).

I really like that most. I like, when there are reasonable, intuitive defaults that cover most of the cases, but then allow the user to decide to make things as complicated as she/he wants. 

If would probably like to see these three functions (at least):

```
catchCondition(f; condition=isempty,  default=missing) = x -> condition(x) ? default : f(x)
catchAny(f; condition=isempty,  default=missing) = x -> condition(x) ? default : try f(x) catch e  default end
catchMissing(f; catchFun=catchCondition) = x -> catchFun(f)(skipmissing(x))
```

The last one would just allow avoiding typing `skipmissing` everytime. Just for convenience, but I think important for easy usage.

I would like to be able to pass the `kwargs` from `catchMissing` also to the other functions, but somehow  `catchMissing(f; catchFun=catchCondition, kwargs...)` did not work (sorry too new to Julia...). 

> Tangential, but I don't really like that the predicate functions we have are positive in the degenerate case rather than the good case. If there were better words for nonmissing/nonempty/nonnan that could be nicer than the current system.

Agree - that's why I'd suggest having "catch" in the name, which makes it more intuitive logically IMO. 


>`whenever(predicate, f, default) = x -> predicate(x) ? default : f(x)`
>It works for isempty, isnothing, ismissing, isnan. I like the positional version more than the kwargs version suggested above. I'm not committed to the name at the moment.

I really like the keyword version with defaults much more. Also I think for the user it is very important to have a consistent design, e.g. here always have `f` at the beginning. This is what makes `tidyverse` so attractive I believe.

**Comment by bkamins** [comment]
Apart from the general solution discussed here, in https://github.com/JuliaData/Missings.jl/pull/143 I propose adding `emptymissing` as @nalimilan since it covers the most common problematic case.

**Comment by nalimilan** [comment]
Thanks @bkamins. Regarding the discussion about more general suggestions, note that this issue is currently in Statistics.jl, but it wouldn't be super logical to add such a general function in this package IMO. It's important to think about where it would go if that function was to be added.

**Comment by bkamins** [comment]
> It's important to think about where it would go if that function was to be added.

I agree. That is why I thought adding `emptymissing` in Missings.jl was a clean choice.

I thought, that once a consensus is reached about the API, a decision can be made where such a function should live (maybe even Base Julia if it is general enough)
